README.md
# dearpygui-map
Map widget for Dear PyGui

## Installation

### Requirements

- Python >= 3.10
- Dear PyGui >= 1.3.1

### Option 1 - pip

You can install this package from pip, with

    pip install dearpygui-map

### Option 2 - Local install from sources

Clone code repository from your local machine, install from there. [Poetry][poetry-install] is required.

    git clone https://github.com/mkouhia/dearpygui-map.git
    cd dearpygui-map
    poetry build
    # take note of the build step output, install package from the dist folder
    pip install dist/PRODUCED_PACKAGE.whl


## Usage

This basic example creates a map widget with defined size, initial center point (latitude, longitude) and initial zoom level. The zoom levels are same as on [tiled maps][tile-zoom-levels], for example 12 could represent a town-sized view. Larger is more zoomed in.

```python
import dearpygui.dearpygui as dpg
import dearpygui_map as dpg_map

dpg.create_context()

with dpg.window(label="Map demo"):
    dpg_map.add_map_widget(
        width=700,
        height=500,
        center=(60.1641, 24.9402),
        zoom_level=12)

dpg.create_viewport(title="Dear PyGui map widget demo", width=800, height=600)
dpg.setup_dearpygui()
dpg.show_viewport()
dpg.start_dearpygui()
dpg.destroy_context()
```

The function `add_map_widget` places a Dear PyGui drawlist with map contents,
return value is the drawlist tag.

By default, OpenStreetMap is used as the map tile source. This can be configured with `add_map_widget` argument `tile_source`, with similar source definition:

```python
OpenStreetMap = TileServer(
    name="OpenStreetMap",
    base_url="http://{subdomain}.tile.openstreetmap.org/{z}/{x}/{y}.png",
    subdomains=["a", "b", "c"],
    max_zoom_level=19,
    license_text="© OpenStreetMap contributors",
    thread_limit=2,
)
```

## Technical details

Tiles are downloaded from the supplier on a background thread.
Whenever map is zoomed or moved, more tiles are downloaded.
The tiles are cached to local storage path in user cache directory - for different platforms, cache directories are:

    Windows:    C:\Users\<username>\AppData\Local\dearpygui_map\Cache
    Mac OS X:   ~/Library/Caches/dearpygui_map
    Other:      ~/.cache/dearpygui_map


### Known issues

As for now, dearpygui-map is in its early design phase and contains some items that may not be acceptable in production environments.
These issues will be addressed later

- Tile download threads are created every time `TileManager.draw_layer` is called.
- Non-visible tiles are not removed from dearpygui
- Zooming and panning leads to wait times before tiles are ready to be shown.

## Development

### Environment

Poetry is required for development (see [installation instructions][poetry-install])

1. Create development environment with `poetry install`
2. Enter environment with `poetry shell`


### Code quality and testing

All code is to be formatted with `black`:

    black dearpygui_map

and code quality checked with `pylint`:

    pylint dearpygui_map

Tests should be written in `pytest`, targeting maximum practical code coverage. Tests are run with:

    pytest

and test coverage checked with

    pytest --cov

Optionally, html test coverage reports can be produced with

    pytest --cov dearpygui_map --cov-report html


### Contributions

Contributions are welcome. Please also see GitHub issues and milestones.


[poetry-install]: https://python-poetry.org/docs/#installation "Poetry: Installation"
[tile-zoom-levels]: https://wiki.openstreetmap.org/wiki/Zoom_levels "Open Street Map wiki: Zoom levels"


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Added
- Pre-download tiles at two tile radius around visible area.

## Changed
- Draw raw textures instead of static textures, use common texture registry.

### Fixed
- Use proper User-Agent http header when downloading tiles.
- Tune error logging output.

## [0.1.1] - 2022-06-13
### Fixed
- Change Python requirement to 3.10, as at least type hinting employs
  features released in 3.10.

## [0.1.0] - 2022-06-13
### Added
- First release: a map widget for Dear PyGui.
- Pan map by dragging mouse middle button.
- Zoom map by mouse scrolling.

[Unreleased]: https://github.com/mkouhia/dearpygui-map/compare/v0.1.1...HEAD
[0.1.1]: https://github.com/mkouhia/dearpygui-map/compare/v0.1.0...v0.1.1
[0.1.0]: https://github.com/mkouhia/dearpygui-map/releases/tag/v0.1.0



tests/test_geo.py
"""Test geographic functions"""

import pytest

from dearpygui_map.geo import Coordinate


def test_coordinate_repr():
    """Reproduction is as defined"""
    coord_x = 0.8
    coord_y = 0.4

    received = repr(Coordinate(coord_x, coord_y))
    expected = f"Coordinate(x={coord_x}, y={coord_y})"

    assert received == expected


def test_coordinate_equality():
    """Two coordinates with same x, y are equal"""
    first = Coordinate(0.4, 0.3)
    second = Coordinate(0.4, 0.3)

    assert first == second


def test_coordinate_inequality():
    """Two coordinates with different x, y are not equal"""
    first = Coordinate(0.4, 0.3)
    second = Coordinate(0.41, 0.3)

    assert first != second


def test_coordinate_tile_xy():
    """Tile x, y coordinates are expected for Helsinki"""
    coord = Coordinate(x=0.5692783333333333, y=0.28948570416212227)

    expected = (2331, 1185)
    received = coord.tile_xy(zoom=12)

    assert received == expected


def test_coordinate_tile_xy_not_floor():
    """Tile x, y coordinates are not floored"""
    coord = Coordinate(x=0.5692783333333333, y=0.28948570416212227)

    expected = (2331.7640533333333, 1185.7334442480528)
    received = coord.tile_xy(zoom=12, floor_=False)

    assert received == expected


def test_coordinate_latlon():
    """Coordinate to lat/lon conversion works as expected"""
    coord = Coordinate(x=0.5692783333333333, y=0.28948570416212227)

    expected = (60.1641, 24.9402)
    received = coord.latlon()

    assert received == pytest.approx(expected, 1e-8)


def test_coordinate_from_latlon():
    """Construction from latitude/longitude works"""
    received = Coordinate.from_latlon(60.1641, 24.9402)
    expected = Coordinate(x=0.5692783333333333, y=0.28948570416212227)

    assert received == expected


invalid_coords = [
    (60.0, -181.0),
    (60.0, 200.0),
    (88.0, 24.0),
    (-86.0, 24.0),
]


@pytest.mark.parametrize("lat, lon", invalid_coords)
def test_coordinate_from_latlon_out_of_bounds(lat, lon):
    """Out-of bounds latitude/longitude results in error"""
    with pytest.raises(ValueError):
        Coordinate.from_latlon(lat, lon)


def test_coordinate_with_screen_offset():
    """Offset calculation works"""
    coord = Coordinate.from_latlon(60.1641, 24.9402)

    received = coord.with_screen_offset(-350, -250, zoom=12)
    expected = Coordinate.from_latlon(60.20677453949039, 24.820037036132817)

    assert received == expected


tests/test_widget.py
"""Map widget tests"""

import itertools

import pytest
from pytest_mock import MockerFixture

from dearpygui_map.geo import Coordinate
from dearpygui_map.tile_source import OpenStreetMap, TileServer, TileSpec
from dearpygui_map.widget import MapWidget, TileManager


@pytest.fixture(name="map_widget")
def fx_map_widget() -> MapWidget:
    """Fixture map widget initially centered in Helsinki"""
    center = (60.1641, 24.9402)
    return MapWidget(700, 500, center, 12, OpenStreetMap)


@pytest.fixture(name="tile_manager")
def fx_tile_manager() -> TileManager:
    """Fixture: tile manager initially centered in Helsinki"""
    center = Coordinate.from_latlon(60.1641, 24.9402)
    origin = center.with_screen_offset(-350, -250, zoom=12)
    return TileManager(origin, 700, 500, 12, OpenStreetMap)


def test_widget_zoom_on_point_origin_changes(map_widget: MapWidget):
    """When zooming, origin changes if canvas_position is not 0,0"""
    origin_0 = map_widget.origin
    map_widget.zoom_on_point((200, 200), 13)

    assert map_widget.origin != origin_0


def test_widget_zoom_on_point_origins_equal(map_widget: MapWidget):
    """When zooming, new origin is same for widget and tile manager"""
    map_widget.zoom_on_point((200, 200), 13)

    assert map_widget.origin == map_widget.tile_manager.origin


def test_widget_get_screen_coordinates(map_widget: MapWidget):
    """Query point on canvas to get coordinate"""
    received = map_widget.get_coordinate(350, 250)
    expected = Coordinate.from_latlon(60.1641, 24.9402)

    assert received == expected


def test_widget_drag_canvas(map_widget: MapWidget, mocker: MockerFixture):
    """Last drag attribute is updated and tile layer is also dragged"""
    tilemgr_cls = mocker.patch("dearpygui_map.widget.TileManager")
    drag = (100, 100)
    # pylint: disable=protected-access
    map_widget._drag_canvas(*drag)

    assert map_widget.last_drag == drag
    assert tilemgr_cls.drag_layer.called_with(*drag)


def test_widget_finish_drag(map_widget: MapWidget, mocker: MockerFixture):
    """Last drag is reset, origin is moved and tile layer is called

    Drag so that previous center point should be found at origin
    """
    tilemgr_cls = mocker.patch("dearpygui_map.widget.TileManager")
    drag = (-350, -250)  # Drag so that
    # pylint: disable=protected-access
    map_widget._drag_canvas(*drag)
    map_widget._finish_drag()

    assert map_widget.last_drag == (0, 0)
    assert map_widget.origin == Coordinate.from_latlon(60.1641, 24.9402)
    assert tilemgr_cls.finish_drag.called_once


def test_center_point_init(tile_manager: TileManager):
    """Check that zoom level and offset are properly set"""
    assert tile_manager.zoom_level == 12
    assert tile_manager.origin_offset == (-596581.5976533333, -303297.7617275015)


def test_center_point_move(tile_manager: TileManager):
    """Check that zoom level and offset are properly set"""
    zoom_level = 13
    center = Coordinate.from_latlon(59.3315, 18.0560)
    origin = center.with_screen_offset(-350, -250, zoom=zoom_level)

    tile_manager.set_origin(origin, zoom_level)

    assert tile_manager.zoom_level == zoom_level
    assert tile_manager.origin_offset == (-1153409.8236444446, -616473.2951275909)


def test_manager_initial_tiles(tile_manager: TileManager):
    """Initial tiles are correct"""
    expected_coord_range = [(2328, 1182), (2335, 1188)]

    # pylint: disable=protected-access
    received = list(tile_manager._required_tiles_for_view())
    expected = _coords_to_specs(
        tile_manager.tile_server,
        expected_coord_range,
        tile_manager.zoom_level,
    )

    assert received == expected


def test_manager_tiles_after_dragging(tile_manager: TileManager):
    """Required tile range changes correctly

    Dragging towards origin (upper left) should reveal more tiles frow
    lower right corner.
    """
    expected_coord_range = [(2330, 1184), (2337, 1190)]

    # Drag to
    tile_manager.drag_layer(-512, -512)

    # pylint: disable=protected-access
    received = list(tile_manager._required_tiles_for_view())
    expected = _coords_to_specs(
        tile_manager.tile_server,
        expected_coord_range,
        tile_manager.zoom_level,
    )

    assert received == expected


def _coords_to_specs(
    tile_server: TileServer, expected_coord_range: list[tuple], zoom: int
) -> list[TileSpec]:
    min_x, min_y = expected_coord_range[0]
    max_x, max_y = expected_coord_range[1]
    tiles = itertools.product(
        range(min_x, max_x + 1),
        range(min_y, max_y + 1),
        (zoom,),
    )
    return list(itertools.starmap(tile_server.to_tile_spec, tiles))


def test_manager_origin_change_after_drag(tile_manager: TileManager):
    """Origin coordinate is correct after dragging"""

    origin_0 = tile_manager.origin
    tile_manager.drag_layer(-512, -512)
    tile_manager.finish_drag()

    received = tile_manager.origin
    expected = origin_0.with_screen_offset(512, 512, zoom=tile_manager.zoom_level)

    assert received == expected


def test_manager_origin_same_after_zoom(tile_manager: TileManager):
    """When zoom changes, origin stays"""
    origin_0 = tile_manager.origin
    tile_manager.set_origin(origin_0, tile_manager.zoom_level + 1)

    assert tile_manager.origin == origin_0


tests/test_tile_source.py
"""Test tile sources"""

from dearpygui_map.tile_source import TileServer, TileSpec


def test_to_tile_spec():
    """Test tile specification generation"""
    tile_x, tile_y, zoom_level = (2331, 1185, 12)
    tile_size = (256, 256)
    base_url = "test_url"
    subdomains = ["foo"]

    test_provider = TileServer("test_name", base_url, subdomains, 1, tile_size)

    expected = TileSpec(tile_x, tile_y, zoom_level, base_url, subdomains, tile_size)
    received = test_provider.to_tile_spec(tile_x, tile_y, zoom_level)

    assert received == expected


dearpygui_map/io.py
"""Tile downloading"""

import logging
import queue
import shutil
import threading
from typing import Callable, Iterable
import urllib.request

from ._version import __version__
from .tile_source import TileSpec


logger = logging.getLogger(__name__)


class TileHandler(threading.Thread):

    """Caching, threaded download manager"""

    def __init__(
        self, tile_specs: Iterable[TileSpec], callback: Callable, thread_count: int = 1
    ) -> None:
        threading.Thread.__init__(self)
        self.task_queue = queue.LifoQueue()
        self.result_queue = queue.Queue()
        self.thread_count = thread_count
        self.threads: list[DownloadThread] = []
        self.callback = callback

        self.add_to_queue(tile_specs)

    def add_to_queue(self, tile_specs: Iterable[TileSpec]) -> None:
        """Find tiles from cache or add tiles to download queue

        Args:
            tile_specs (Iterable[TileSpec]): tiles to be downloaded
        """
        for tile_spec in tile_specs:
            if tile_spec.local_storage_path.exists():
                self.result_queue.put(tile_spec)
            else:
                self.task_queue.put(tile_spec)

    def run(self):
        """Run download"""
        for _ in range(self.thread_count):
            thread = DownloadThread(self.task_queue, self.result_queue)
            thread.start()
            self.threads.append(thread)
        while not (
            self.result_queue.empty() and sum([t.is_alive() for t in self.threads]) == 0
        ):
            result: TileSpec = self.result_queue.get()
            try:
                self.callback(result)
            except Exception as exc:  # pylint: disable=broad-except
                logger.error(exc)
            finally:
                self.result_queue.task_done()


class DownloadThread(threading.Thread):

    """Threaded downloader"""

    def __init__(self, task_queue: queue.Queue, result_queue: queue.Queue):
        threading.Thread.__init__(self)
        self.task_queue = task_queue
        self.result_queue = result_queue

        self.user_agent = f"dearpygui-map/{__version__}"

    def run(self):
        """Run download"""
        # TODO find out a way to handle failures better
        while not self.task_queue.empty():
            tile_spec: TileSpec = self.task_queue.get()
            req = urllib.request.Request(
                tile_spec.download_url, headers={"User-Agent": self.user_agent}
            )
            try:
                tile_path = tile_spec.local_storage_path
                tile_path.parent.mkdir(parents=True, exist_ok=True)
                with urllib.request.urlopen(req) as response, open(
                    tile_path, "wb"
                ) as local_file:
                    # TODO add error handling etc.
                    shutil.copyfileobj(response, local_file)
                self.result_queue.put(tile_spec)
            except urllib.error.HTTPError as exc:
                logger.error("%s, url: %s", exc, tile_spec.download_url, exc_info=False)

            self.task_queue.task_done()
            # TODO need to store coordinates alongside with image


dearpygui_map/geo.py
"""Geography calculation"""

import math


class Coordinate:

    """Internal coordinate representation

    Attributes:
        x (float): x dimension, display units
        y (float): y dimension, display units
    """

    MAX_LATITUDE = 85.0511287798

    def __init__(self, x: float, y: float) -> None:  # pylint: disable=invalid-name
        # pylint: disable=invalid-name
        self.x = x
        self.y = y

    def __repr__(self) -> str:
        return f"Coordinate(x={self.x!r}, y={self.y!r})"

    def __eq__(self, other: "Coordinate") -> bool:
        return self.x == other.x and self.y == other.y

    @staticmethod
    def _lon_to_x(lon: float) -> float:
        if lon > 180 or lon < -180:
            raise ValueError(f"Longitude must be -180...180, {lon=}")
        return (lon + 180) / 360

    @classmethod
    def _lat_to_y(cls, lat: float) -> float:
        if lat > cls.MAX_LATITUDE or lat < -cls.MAX_LATITUDE:
            val_ = cls.MAX_LATITUDE
            raise ValueError(f"Latitude must be -{val_}...{val_}: {lat=}")

        lat_rad = math.radians(lat)
        return (1.0 - math.asinh(math.tan(lat_rad)) / math.pi) / 2.0

    def tile_xy(self, zoom: int, floor_: bool = True) -> tuple[int, int]:
        """Calculate tile x, y coordinates

        If coordinate falls out of projection (x, y < 0 or > max value),
        just truncate to min/max tile

        Args:
            zoom (int): Zoom level
            floor_ (bool): Truncate coordinates to integer tile numbers

        Returns:
            tuple[int, int]: Tile x, y coordinates
        """
        zoom_scale = math.pow(2, zoom)
        tile_x = self.x * zoom_scale
        tile_y = self.y * zoom_scale

        tile_x = min(max(tile_x, 0), zoom_scale - 1)
        tile_y = min(max(tile_y, 0), zoom_scale - 1)

        if floor_:
            tile_x = math.floor(tile_x)
            tile_y = math.floor(tile_y)

        return (tile_x, tile_y)

    def latlon(self) -> tuple[float, float]:
        """Convert internal coordinates to lat, lon

        Returns:
            tuple[float, float]: latitude, longitude
        """
        lon_deg = self.x * 360.0 - 180.0
        lat_rad = math.atan(math.sinh(math.pi * (1 - 2 * self.y)))
        lat_deg = math.degrees(lat_rad)
        return (lat_deg, lon_deg)

    @classmethod
    def from_latlon(cls, latitude: float, longitude: float) -> "Coordinate":
        """Construct internal coordinate representation from lat /lon

        Args:
            latitude (float): Latitude, degrees
            longitude (float): Longitude, degrees

        Raises:
            ValueError: Latitude or longitude is out of bounds: latitude
              must be within -85.0511287798 ... 85.0511287798 (map tile)
              restriction and longitude within -180 ... 180.

        Returns:
            Coordinate: Internal coordinate instance
        """
        coord_x = cls._lon_to_x(longitude)
        coord_y = cls._lat_to_y(latitude)
        return cls(coord_x, coord_y)

    def with_screen_offset(
        self,
        delta_x: float,
        delta_y: float,
        *,
        zoom: int,
        resolution: tuple[int, int] = (256, 256),
    ) -> "Coordinate":
        """Return new coordinate with offset location

        Args:
            delta_x (float): Difference in x coordinate, screen pixels
            delta_y (float): Difference in y coordinate, screen pixels
            zoom (int): Zoom level
            resolution (tuple[int, int]): Pixel resolution per zoom unit

        Returns:
            Coordinate: New coordinate with offset
        """
        dim_tiles = math.pow(2, zoom)
        new_x = self.x + delta_x / (dim_tiles * resolution[0])
        new_y = self.y + delta_y / (dim_tiles * resolution[1])
        return Coordinate(new_x, new_y)


dearpygui_map/tile_source.py
"""Map tile sources"""

from dataclasses import dataclass
from urllib.parse import urlparse, unquote
from pathlib import Path

from dearpygui_map.util import user_cache_dir


@dataclass
class TileServer:

    """Map tile source"""

    name: str
    base_url: str
    subdomains: list[str]
    max_zoom_level: int
    license_text: str
    thread_limit: int = 1
    tile_size: tuple[int, int] = (256, 256)

    def to_tile_spec(self, tile_x: int, tile_y: int, zoom_level: int) -> "TileSpec":
        """Get tile specification for x, y, z coordinate tuple

        Args:
            tile_x (int): Tile x coordinate
            tile_y (int): Tile y coordinate
            zoom_level (int): Zoom level

        Returns:
            TileSpec: Tile specification
        """
        return TileSpec(
            tile_x=tile_x,
            tile_y=tile_y,
            zoom_level=zoom_level,
            tile_size=self.tile_size,
            base_url=self.base_url,
            subdomains=self.subdomains,
        )


@dataclass
class TileSpec:

    """Specification of a tile on server"""

    tile_x: int
    tile_y: int
    zoom_level: int
    base_url: str
    subdomains: list[str]
    tile_size: tuple[int, int] = (256, 256)

    @property
    def download_url(self):
        """Get download url for tile"""
        return self.base_url.format(
            subdomain=self.subdomains[0],
            x=self.tile_x,
            y=self.tile_y,
            z=self.zoom_level,
        )

    @property
    def local_storage_path(self) -> Path:
        """Get file location on local device"""
        cache_root = user_cache_dir()
        local_url = self.base_url.replace("{subdomain}.", "").format(
            x=self.tile_x,
            y=self.tile_y,
            z=self.zoom_level,
        )
        components = urlparse(local_url)
        Path(unquote(components.path))
        return Path(
            cache_root, components.netloc, *Path(unquote(components.path)).parts[1:]
        )

    def canvas_coordinates(self) -> tuple[int, int]:
        """Calculate canvas coordinates for tile

        Returns:
            tuple[int, int]: position (x, y)
        """
        return (
            self.tile_x * self.tile_size[0],
            self.tile_y * self.tile_size[1],
        )


OpenStreetMap = TileServer(
    name="OpenStreetMap",
    base_url="http://{subdomain}.tile.openstreetmap.org/{z}/{x}/{y}.png",
    subdomains=["a", "b", "c"],
    max_zoom_level=19,
    license_text="© OpenStreetMap contributors",
    thread_limit=2,
)
